### PR TITLE
VAD audio chunking

### PR DESCRIPTION
**Pull request by jkrukowski**
This PR introduces audio chunking with VAD. The VAD is used to detect speech segments in the audio file and then the audio is split into chunks based on the detected speech segments (and padded with zeros to match the 30sec length). Chunks are then processed in a batch resulting in a significant speedup.

Some benchmarks (on my mac book air m1):

### Audio file 12:16 length

* with VAD:

```
38.16s user 5.86s system 470% cpu 9.349 total
```

* without VAD:

```
33.25s user 3.55s system 132% cpu 27.678 total
```

### Audio file 40:26 length

* with VAD:

```
126.54s user 18.41s system 500% cpu 28.952 total
```

* without VAD:

```
96.55s user 10.47s system 133% cpu 1:20.08 total
```

To use it in WhisperKitCLI the user has to pass the `chunking-strategy` flag:

```
swift run -c release whisperkit-cli transcribe --audio-path /path/to/audio.wav --chunking-strategy vad
```

### CODE REVIEW

**Comment by atiorh** [comment]
Great work @jkrukowski! Did you see the Cut and Merge strategy in https://github.com/m-bain/whisperX?

Screenshot 2024-05-06 at 10 10 29 AM

If we don't attempt to pack short segments into 30 seconds like above before padding, the worst case performance might regress below baseline (e.g. padding ~1-3s chunks to 30s a lot). Let me know if you think Cut and Merge is an extension we should leave as future work or bundle here :)

**Comment by atiorh** [comment]
@Abhinay1997's Do you mind rebasing on top of this PR so we can add a WER check (w/ and w/o VAD-based chunking) on your long audio test sample? 🙏

**Comment by Abhinay1997** [comment]
Hey @atiorh ! No worries, I'll do that by tomorrow. Want to make sure there are no bugs/crash prone code in my PR.

**Comment by jkrukowski** [comment]
> Great work @jkrukowski! Did you see the Cut and Merge strategy in https://github.com/m-bain/whisperX?
> 
> Screenshot 2024-05-06 at 10 10 29 AM
> If we don't attempt to pack short segments into 30 seconds like above before padding, the worst case performance might regress below baseline (e.g. padding ~1-3s chunks to 30s a lot). Let me know if you think Cut and Merge is an extension we should leave as future work or bundle here :)
> 
> Edit: Cut and Merge will also mean some additional bookkeeping to adjust word-level timetamps post-inference.

I'd leave it as a future work if possible. After talking to @ZachNagengast I took a bit different approach here -- using VAD I'm trying to find the best cut off point in the 2nd half of 30sec audio chunk. So there is no risk of having a bunch of small segments padded with zeros (because the segment will contain at least 15 sec of the original audio chunk). Having said that I think that cut and merge is a better (but more complicated) approach

**Comment by atiorh** [comment]
> > Great work @jkrukowski! Did you see the Cut and Merge strategy in https://github.com/m-bain/whisperX?
> > If we don't attempt to pack short segments into 30 seconds like above before padding, the worst case performance might regress below baseline (e.g. padding ~1-3s chunks to 30s a lot). Let me know if you think Cut and Merge is an extension we should leave as future work or bundle here :)
> > Edit: Cut and Merge will also mean some additional bookkeeping to adjust word-level timetamps post-inference.
> 
> I'd leave it as a future work if possible. After talking to @ZachNagengast the other day I took a bit different approach here -- using VAD I'm trying to find the best cut off point in the 2nd half of 30sec audio chunk. So there is no risk of having a bunch of small segments padded with zeros (because the segment will contain at least 15 sec of the original audio). Having said that I think that cut and merge is a better (but more complicated) approach

Makes sense, this is great.

File: Sources/WhisperKit/Core/EnergyVAD.swift [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+        self.frameShift = frameShift * sampleRate / 1000
+        self.energyThreshold = energyThreshold
+    }
+
+    func voiceActivity(in waveform: [Float]) -> [Bool] {
+        let count = (waveform.count - frameLength + frameShift) / frameShift
+        if count < 0 {
+            return []
+        }
+        return (0..<count).map { vDSP.sumOfSquares(waveform[$0 * frameShift..<($0 * frameShift + frameLength)]) > energyThreshold }
```
Comment (ZachNagengast) [inline Sources/WhisperKit/Core/EnergyVAD.swift:38]: I'd pull this into a helper function so it's a little easier to read. There is a similar section of code doing this that should move over https://github.com/argmaxinc/WhisperKit/blob/d6f50da1dc0b69fc53bb7d4a14f3972cc1353174/Sources/WhisperKit/Core/AudioProcessor.swift#L287-L321

File: Sources/WhisperKit/Core/EnergyVAD.swift [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+        energyThreshold: Float = 0.05
+    ) {
+        self.sampleRate = sampleRate
+        // Compute frame length and frame shift in number of samples (not milliseconds)
+        self.frameLength = frameLength * sampleRate / 1000
+        self.frameShift = frameShift * sampleRate / 1000
+        self.energyThreshold = energyThreshold
+    }
+
+    func voiceActivity(in waveform: [Float]) -> [Bool] {
```
[the remark this answers is not in the document]
Comment (jkrukowski) [reply Sources/WhisperKit/Core/EnergyVAD.swift:34]: Added `calculateNonSilentChunks` method in `AudioProcessor` which is backed by `EnergyVAD`, this way we can keep `EnergyVAD` class internal

File: Sources/WhisperKit/Core/EnergyVAD.swift [hunk 3]
```diff
@@ -0,0 +1,76 @@
+//  For licensing see accompanying LICENSE.md file.
+//  Copyright © 2024 Argmax, Inc. All rights reserved.
+
+import Foundation
+import Accelerate
+
+/// Voice activity detection based on energy threshold
+final class EnergyVAD {
```
[the remark this answers is not in the document]
Comment (jkrukowski) [reply Sources/WhisperKit/Core/EnergyVAD.swift:9]: for now I decided to move this other VAD code in our repo to `isVoiceDetected` method in `AudioProcessor`. This way we can keep `EnergyVAD` internal till the full public interface of this class is ready

File: Sources/WhisperKit/Core/Utils.swift [hunk 4]
```diff
@@ -506,6 +585,22 @@ public func mergeTranscriptionResults(_ results: [TranscriptionResult?], confirm
     )
 }
 
+public func updateSegmentTimings(segment: TranscriptionSegment, seekTime: Float) -> TranscriptionSegment {
```
Comment (ZachNagengast) [inline Sources/WhisperKit/Core/Utils.swift:588]:
```suggestion
@available(macOS 13, iOS 16, watchOS 10, visionOS 1, *)
public func updateSegmentTimings(segment: TranscriptionSegment, seekTime: Float) -> TranscriptionSegment {
```

**Comment by ZachNagengast** [comment]
Here's a recording of the example app running chunking about 4x faster with minimal WER loss 🚀 

https://github.com/argmaxinc/WhisperKit/assets/1981179/fe403ceb-b752-4396-ab7a-905eb3351c40

**Review by ZachNagengast — APPROVED** [review]
Approved and good to go. Excellent work on this @jkrukowski this will be a massive improvement across the board for large workloads, and even helps with streaming if the buffer gets backed up by >30s

**Pull request merged**